parlai/agents/examples/seq2seq.py

```diff
@@ -0,0 +1,202 @@
+    """
+    Example encoder.
+
+    Pay particular attention to the ``forward`` output.
+    """
+
+    def __init__(self, embeddings, hidden_size):
+        """
+        Initialization.
+
+        Arguments here can be used to provide hyperparameters.
+        """
+        # must call super on all nn.Modules.
+        super().__init__()
+
+        _vocab_size, esz = embeddings.weight.shape
+        self.embeddings = embeddings
+        self.lstm = nn.LSTM(
+            input_size=esz, hidden_size=hidden_size, num_layers=1, batch_first=True
+        )
+
+    def forward(self, input_tokens):
+        """
+        Perform the forward pass for the encoder.
+
+        Input *must* be input_tokens, which are the context tokens given
+        as a matrix of lookup IDs.
+
+        :param input_tokens:
+            Input tokens as a bsz x seqlen LongTensor.
+            Likely will contain padding.
+
+        :return:
+            You can return anything you like; it is will be passed verbatim
+            into the decoder for conditioning. However, it should be something
+            you can easily manipulate in ``reorder_encoder_states``.
+            This particular implementation returns the hidden and cell states from the
+            LSTM.
+        """
+        embedded = self.embeddings(input_tokens)
+        _output, hidden = self.lstm(embedded)
+        return hidden
+
+
+class Decoder(nn.Module):
+    """
+    Basic example Decoder.
+
+    Pay particular note to the ``forward``.
+    """
+
+    def __init__(self, embeddings, hidden_size):
+        """
+        Initialization.
+
+        Arguments here can be used to provide hyperparameters.
+        """
+        super().__init__()
+        _vocab_size, self.esz = embeddings.weight.shape
+        self.embeddings = embeddings
+        self.lstm = nn.LSTM(
+            input_size=self.esz, hidden_size=hidden_size, num_layers=1, batch_first=True
+        )
+
+    def forward(self, input, encoder_state, incr_state=None):
+        """
+        Run forward pass.
+
+        :param input:
+            The currently generated tokens from the decoder.
+        :param encoder_state:
+            The output from the encoder module.
+        :parm incr_state:
+            The previous hidden state of the decoder.
+        """
+        embedded = self.embeddings(input)
+        if incr_state is None:
+            # this is our very first call. We want to seed the LSTM with the
+            # hidden state of the decoder
+            state = encoder_state
+        else:
+            # We've generated some tokens already, so we can reuse the existing
+            # decoder state
+            state = incr_state
+
+        # get the new output and decoder incremental state
+        output, incr_state = self.lstm(embedded, state)
+
+        return output, incr_state
+
+
+class ExampleModel(tga.TorchGeneratorModel):
+    """
+    ExampleModel implements the final components of TorchGeneratorModel.
```

Comment:
a bit confusing, doesn't this implement the entire model?

Reply:
yes - reworded